### PR TITLE
feat(rust, python): Return datetime for mean/median of Date column

### LINKED ISSUE #13598: `mean` and `median` for `pl.Date` should return a datetime
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from datetime import date
import polars as pl

s = pl.Series([date(2024, 1, 1), date(2024, 1, 2)])
print("mean:")
print(s.mean())
print(s.dt.mean())
print(s.to_frame().select(pl.all().mean()))

print("\nmedian:")
print(s.median())
print(s.dt.median())
print(s.to_frame().select(pl.all().median()))
```

### Log output

```shell
mean:
19723.5
2024-01-01
shape: (1, 1)
┌──────┐
│      │
│ ---  │
│ date │
╞══════╡
│ null │
└──────┘

median:
19723.5
2024-01-01
polars.exceptions.InvalidOperationError: `median` operation not supported for dtype `date`
```

### Issue description

`dt.mean()` returns a date and `mean()` returns a float or null, depending on context, when the input has dtype `pl.Date`. Median similarly has inconsistent behavior.

Using the analogy for integers and real numbers, the mean and median of dates should be a datetime, similar to how the mean and median of an integers is a float:

```
mean(1, 2) = 1.5
mean(2024-01-01, 2024-01-02) = 2024-01-01 12:00:00
```

### Expected behavior

Returns a datetime that is at the average (mean or median) time point, and is inconsistent across uses of `mean` and `median`.

Implementing this for the remaining temporal types (in a future issue) will enable the deprecation of `.dt.mean` and `.dt.median`, as those will become redundant with `series.mean` and `series.median` when the datatype is temporal.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.3
Index type:           UInt32
Platform:             Windows-10-10.0.22621-SP0
Python:               3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          3.0.0
connectorx:           0.3.2
deltalake:            0.15.1
fsspec:               2023.12.2
gevent:               23.9.1
hvplot:               0.9.1
matplotlib:           3.8.2
numpy:                1.26.3
openpyxl:             3.1.2
pandas:               2.1.4
pyarrow:              14.0.2
pydantic:             2.5.3
pyiceberg:            0.5.1
pyxlsb:               1.0.10
sqlalchemy:           2.0.25
xlsx2csv:             0.8.1
xlsxwriter:           3.1.9
```

</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #13598. Part of overall #13599.

This treats dates and datetimes as analagous to integers/floats: the mean of ints is a float, and now the mean of dates is a datetime:

```
mean(1, 2) = 1.5
mean(2023-01-01, 2023-01-02) = 2023-01-01 12:00.
```

This is the last temporal mean/median PR, which means a lot of custom logic for the different temporal types has been collapsed/simplified. If this is accepted, we can deprecate `dt.mean` and `dt.median`, as mean/median for all temporals have now been implemented, they are now functionally equivalent to `series.mean` and `series.median` for all temporal dtypes. I'll do a followup on `std`/`var`/`quantile`.

This PR is almost identical to #13492, but that was pretty old/stale and cluttered, so easier to just make a new one. @ritchie46 to get streaming to work, I had to update the `MeanAgg` strutc with a flag to indicat the Date -> Datetime conversion. AFAIK Date is the only dtype that requires a value conversion for mean now, so I don't think we need a more general implementation, but you may think otherwise.

### CODE REVIEW

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 1]
```diff
@@ -166,29 +139,66 @@ impl Series {
...
+        }
 
         match self.dtype() {
             Boolean => self.cast(&Float64).unwrap().agg_mean(groups),
             Float32 => SeriesWrap(self.f32().unwrap().clone()).agg_mean(groups),
             Float64 => SeriesWrap(self.f64().unwrap().clone()).agg_mean(groups),
             dt if dt.is_numeric() => apply_method_physical_integer!(self, agg_mean, groups),
+            #[cfg(feature = "dtype-date")]
+            Date => temporal_mean(
+                &(self.cast(&Int64).unwrap() * (MS_IN_DAY as f64)),
```
Comment (ritchie46) [inline crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:159]: Can we cast this after the aggregation. That should be much cheaper.

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 2]
```diff
@@ -166,29 +139,66 @@ impl Series {
...
+        }
+
+        match self.dtype() {
+            Boolean => self.cast(&Float64).unwrap().agg_median(groups),
+            Float32 => SeriesWrap(self.f32().unwrap().clone()).agg_median(groups),
+            Float64 => SeriesWrap(self.f64().unwrap().clone()).agg_median(groups),
+            dt if dt.is_numeric() => apply_method_physical_integer!(self, agg_median, groups),
+            #[cfg(feature = "dtype-date")]
+            Date => temporal_median(
+                &(self.cast(&Int64).unwrap() * (MS_IN_DAY as f64)),
```
Comment (ritchie46) [inline crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:192]: Idem

File: crates/polars-pipe/src/executors/sinks/group_by/aggregates/interface.rs [hunk 3]
```diff
@@ -54,6 +54,7 @@ pub(crate) enum AggregateFunction {
     SumU64(SumAgg<u64>),
     SumI32(SumAgg<i32>),
     SumI64(SumAgg<i64>),
+    MeanDate(MeanAgg<i64>),
```
Comment (ritchie46) [inline crates/polars-pipe/src/executors/sinks/group_by/aggregates/interface.rs:57]: Can we remove this for now. We are redesigning the streaming engine and don't want to add stuff atm.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

**Comment by ritchie46** [comment]
Thank you @mcrumiller. I have left a few comments. I think we should remove the streaming engine for now. That's something I want to redesign.

File: crates/polars-pipe/src/executors/sinks/group_by/aggregates/interface.rs (hunk 3 above)
Comment (mcrumiller) [reply crates/polars-pipe/src/executors/sinks/group_by/aggregates/interface.rs:57]: Yep, it all felt ugly to me anyway.

**Comment by mcrumiller** [comment]
@ritchie46 removing the streaming code ends up breaking `date.mean` in streaming, since the streaming engine uses different logic to calculate the mean, but still uses the schema logic to cast the result to a `Datetime`. The result is that the date's mean value gets directly cast to an `i64` with no conversion, so the streaming mean always shows something like 1970-01-01 00:42:63 or some random time.

Should I try to find a workaround, or should I remove the date mean/median from the streaming tests?

**Comment by ritchie46** [comment]
After thinking of it a bit. We should remove the physical implementations (we could panic there) and do all the casts on the logical side. During type-coercion we can set the proper casts. This will ensure it works on all physical engines. And we panic on the physical side to ensure we don't hit those paths

**Comment by mcrumiller** [comment]
> After thinking of it a bit. We should remove the physical implementations (we could panic there) and do all the casts on the logical side. During type-coercion we can set the proper casts. This will ensure it works on all physical engines. And we panic on the physical side to ensure we don't hit those paths

Thanks @ritchie46. I may need a little help here though. Are you referring to logical and physical dtypes (I think you are), or logical plan versus physical plan? Also, this sort of sounds like it might be better off as a separate PR, do you agree?

**Comment by mcrumiller** [comment]
@ritchie46 any proposal for how to proceed with this one? I'm sort of waiting on the streaming engine rewrite. Should we merge and just have date mean broken for streaming until that comes out? Or alternatively, keep my prior "hack" to get it working on the current engine? Or maybe I missed something from your comment that means I can resolve this immediately. I've been busy and let this slip.

**Pull request closed without merging**